File: Src2/Icm.Core/CommandLine/NamedParameter.vb
Imports res = My.Resources.CommandLineTools

Namespace Icm.CommandLineTools

    ''' <summary>
    '''  Command line option.
    ''' </summary>
    ''' <remarks></remarks>
    Public Class NamedParameter

        Private _shortName As String
        Private _longName As String
        Private ReadOnly _subarguments As List(Of SubArgument)
        Private ReadOnly _isRequired As Boolean
        Private _unlimitedSubargs As Boolean

        Public Sub New(
                ByVal sShortName As String,
                ByVal sLongName As String,
                ByVal sDescription As String,
                ByVal bRequired As Boolean,
                ByVal ParamArray subargs() As SubArgument)

            Debug.Assert(sShortName.Length <= sLongName.Length)
            ShortName = sShortName
            LongName = sLongName
            Description = sDescription
            _isRequired = bRequired
            _subarguments = New List(Of SubArgument)
            If subargs IsNot Nothing Then
                For Each subarg In subargs
                    AddSubArgument(subarg)
                Next
            End If
            MustReset = False
        End Sub

        ''' <summary>
        ''' If this property is True, it indicates that we must eliminate all the present
        ''' subarguments.
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Property MustReset As Boolean

        ''' <summary>
        '''  If true, the presence of this option in the command line makes the processor to
        ''' ignore the main arguments because they are unneeded.
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Property MakesMainArgumentsIrrelevant() As Boolean

        ''' <summary>
        ''' Does this option accept an unlimited list of subarguments?
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        ReadOnly Property AcceptsUnlimitedSubargs() As Boolean
            Get
                Return _unlimitedSubargs
            End Get
        End Property

        ReadOnly Property IsRequired As Boolean
            Get
                Return _isRequired
            End Get
        End Property

        ''' <summary>
        '''  Short option (for example -h)
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Property ShortName() As String
            Get
                Return _shortName
            End Get
            Set(ByVal Value As String)
                Debug.Assert(Value.Length >= 1)
                _shortName = Value
            End Set
        End Property

        ''' <summary>
        '''  Long option (for example --help)
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Property LongName() As String
            Get
                Return _longName
            End Get
            Set(ByVal Value As String)
                Debug.Assert(Value.Length > 1)
                _longName = Value
            End Set
        End Property

        ''' <summary>
        '''  Description of this option.
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Property Description() As String

        ReadOnly Property SubArguments() As List(Of SubArgument)
            Get
                Return _subarguments
            End Get
        End Property

        ''' <summary>
        '''  Adds a subargument.
        ''' </summary>
        ''' <param name="sa">Subargument to be added</param>
        ''' <remarks></remarks>
        Public Sub AddSubArgument(ByVal sa As SubArgument)
            ' require
            '   p /= Void
            '   parameters.Last.isOptional implies p.isOptional
            '   not parameters.Last.isList
            Debug.Assert(Not sa Is Nothing, res.S_ERR_VOID_SUBARGUMENT)

            If _subarguments.Count > 0 Then
                Dim lastsa As SubArgument = CType(_subarguments(_subarguments.Count - 1), SubArgument)
                Debug.Assert(Not lastsa.IsOptional OrElse sa.IsOptional, String.Format(res.S_ERR_NON_OPTIONAL_AFTER_OPTIONAL, sa.Name, lastsa.Name))
                Debug.Assert(Not lastsa.IsList, res.S_ERR_SUBARG_AFTER_LIST)
            End If
            If sa.IsList Then
                _unlimitedSubargs = True
            End If
            _subarguments.Add(sa)
        End Sub

    End Class
End Namespace

File: Src/Icm.Core/CommandLine/NamedParameter.cs

using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using res = My.Resources.CommandLineTools;

namespace Icm.CommandLineTools
{

	/// <summary>
	///  Command line option.
	/// </summary>
	/// <remarks></remarks>
	public class NamedParameter
	{

		private string _shortName;
		private string _longName;
		private readonly List<SubArgument> _subarguments;
		private readonly bool _isRequired;

		private bool _unlimitedSubargs;

		public NamedParameter(string sShortName, string sLongName, string sDescription, bool bRequired, params SubArgument[] subargs)
		{
			Debug.Assert(sShortName.Length <= sLongName.Length);
			ShortName = sShortName;
			LongName = sLongName;
			Description = sDescription;
			_isRequired = bRequired;
			_subarguments = new List<SubArgument>();
			if (subargs != null) {
				foreach (void subarg_loopVariable in subargs) {
					subarg = subarg_loopVariable;
					AddSubArgument(subarg);
				}
			}
			MustReset = false;
		}

		/// <summary>
		/// If this property is True, it indicates that we must eliminate all the present
		/// subarguments.
		/// </summary>
		/// <value></value>
		/// <returns></returns>
		/// <remarks></remarks>
		public bool MustReset { get; set; }

		/// <summary>
		///  If true, the presence of this option in the command line makes the processor to
		/// ignore the main arguments because they are unneeded.
		/// </summary>
		/// <value></value>
		/// <returns></returns>
		/// <remarks></remarks>
		public bool MakesMainArgumentsIrrelevant { get; set; }

		/// <summary>
		/// Does this option accept an unlimited list of subarguments?
		/// </summary>
		/// <value></value>
		/// <returns></returns>
		/// <remarks></remarks>
		public bool AcceptsUnlimitedSubargs {
			get { return _unlimitedSubargs; }
		}

		public bool IsRequired {
			get { return _isRequired; }
		}

		/// <summary>
		///  Short option (for example -h)
		/// </summary>
		/// <value></value>
		/// <returns></returns>
		/// <remarks></remarks>
		public string ShortName {
			get { return _shortName; }
			set {
				Debug.Assert(value.Length >= 1);
				_shortName = value;
			}
		}

		/// <summary>
		///  Long option (for example --help)
		/// </summary>
		/// <value></value>
		/// <returns></returns>
		/// <remarks></remarks>
		public string LongName {
			get { return _longName; }
			set {
				Debug.Assert(value.Length > 1);
				_longName = value;
			}
		}

		/// <summary>
		///  Description of this option.
		/// </summary>
		/// <value></value>
		/// <returns></returns>
		/// <remarks></remarks>
		public string Description { get; set; }

		public List<SubArgument> SubArguments {
			get { return _subarguments; }
		}

		/// <summary>
		///  Adds a subargument.
		/// </summary>
		/// <param name="sa">Subargument to be added</param>
		/// <remarks></remarks>
		public void AddSubArgument(SubArgument sa)
		{
			// require
			//   p /= Void
			//   parameters.Last.isOptional implies p.isOptional
			//   not parameters.Last.isList
			Debug.Assert((sa != null), res.S_ERR_VOID_SUBARGUMENT);

			if (_subarguments.Count > 0) {
				SubArgument lastsa = (SubArgument)_subarguments(_subarguments.Count - 1);
				Debug.Assert(!lastsa.IsOptional || sa.IsOptional, string.Format(res.S_ERR_NON_OPTIONAL_AFTER_OPTIONAL, sa.Name, lastsa.Name));
				Debug.Assert(!lastsa.IsList, res.S_ERR_SUBARG_AFTER_LIST);
			}
			if (sa.IsList) {
				_unlimitedSubargs = true;
			}
			_subarguments.Add(sa);
		}

	}
}

//=======================================================
//Service provided by Telerik (www.telerik.com)
//Conversion powered by NRefactory.
//Twitter: @telerik
//Facebook: facebook.com/telerik
//=======================================================
